src/Nethermind/Nethermind.Evm/EvmStack.cs

```diff
@@ -181,51 +185,53 @@ public void PushUInt256(in UInt256 value)
             Span<byte> word = _bytes.Slice(Head * 32, 32);
             ref byte bytes = ref word[0];
 
-            ulong u3 = value.u3;
-            ulong u2 = value.u2;
-            ulong u1 = value.u1;
-            ulong u0 = value.u0;
-
+            ulong u3, u2, u1, u0;
             if (BitConverter.IsLittleEndian)
             {
-                u3 = BinaryPrimitives.ReverseEndianness(u3);
-                u2 = BinaryPrimitives.ReverseEndianness(u2);
-                u1 = BinaryPrimitives.ReverseEndianness(u1);
-                u0 = BinaryPrimitives.ReverseEndianness(u0);
+                u3 = BinaryPrimitives.ReverseEndianness(value.u3);
+                u2 = BinaryPrimitives.ReverseEndianness(value.u2);
+                u1 = BinaryPrimitives.ReverseEndianness(value.u1);
+                u0 = BinaryPrimitives.ReverseEndianness(value.u0);
+            }
+            else
+            {
+                u3 = value.u3;
+                u2 = value.u2;
+                u1 = value.u1;
+                u0 = value.u0;
             }
 
-            Unsafe.WriteUnaligned(ref bytes, u3);
-            Unsafe.WriteUnaligned(ref Unsafe.Add(ref bytes, sizeof(ulong)), u2);
-            Unsafe.WriteUnaligned(ref Unsafe.Add(ref bytes, 2 * sizeof(ulong)), u1);
-            Unsafe.WriteUnaligned(ref Unsafe.Add(ref bytes, 3 * sizeof(ulong)), u0);
+            Unsafe.WriteUnaligned(ref bytes, Vector256.Create(u3, u2, u1, u0));
 
             if (_tracer.IsTracingInstructions) _tracer.ReportStackPush(word);
 
             if (++Head >= MaxStackSize)
             {
                 Metrics.EvmExceptions++;
-                throw new EvmStackOverflowException();
+                ThrowEvmStackOverflowException();
             }
         }
 
         public void PushSignedInt256(in Int256.Int256 value)
         {
-            PushUInt256((UInt256)value);
+            // tail call into UInt256
```

Comment:
QQ: Can you elaborate what you mean by "tail call" here? Does .net have some specific optimication in this case?

Reply:
No stack to clean up; return type and parameters match; so in the asm these method just turns into a single jump to the other methods
```asm
; Method Nethermind.Evm.EvmStack:PushSignedInt256(byref):this
G_M000_IG01:

G_M000_IG02:
       FF25D23F1400         tail.jmp [Nethermind.Evm.EvmStack:PushUInt256(byref):this]
; Total bytes of code: 6
```
```asm
; Method Nethermind.Evm.EvmStack:PopSignedInt256(byref):this
G_M000_IG01:

G_M000_IG02:
       FF2532401400         tail.jmp [Nethermind.Evm.EvmStack:PopUInt256(byref):this]
; Total bytes of code: 6

```